infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -203,41 +206,42 @@ private void processOpaqueEntry(List<SMREntry> smrEntries, UUID shadowStreamUuid
-        TxBuilder txBuilder = logReplicationMetadataManager.getTxBuilder();
-        CorfuStoreMetadata.Timestamp timestamp = processOpaqueEntryShadowStream(txBuilder, smrEntries, currentSeqNum, shadowStreamUuid);
+        CorfuStoreMetadata.Timestamp timestamp;
+
+        try (TxnContext txn = logReplicationMetadataManager.getTxnContext()) {
+            processOpaqueEntryShadowStream(txn, smrEntries, currentSeqNum, shadowStreamUuid);
+            timestamp = txn.commit();
+        }
 
-        try {
-            if (!snapshotSyncStartMarker.isPresent()) {
-                logReplicationMetadataManager.setSnapshotSyncStartMarker(snapshotSyncId, timestamp, txBuilder);
+        if (!snapshotSyncStartMarker.isPresent()) {
+            try (TxnContext txn = logReplicationMetadataManager.getTxnContext()) {
+                logReplicationMetadataManager.setSnapshotSyncStartMarker(txn, snapshotSyncId, timestamp);
                 snapshotSyncStartMarker = Optional.of(new SnapshotSyncStartMarker(snapshotSyncId, timestamp.getSequence()));
             }
-            txBuilder.commit(timestamp);
-        } catch (Exception e) {
-            log.warn("Caught an exception ", e);
-            throw e;
         }
         log.debug("Process entries total={}, set sequence number {}", smrEntries.size(), currentSeqNum);
     }
 
-    private CorfuStoreMetadata.Timestamp processOpaqueEntryShadowStream(TxBuilder txBuilder, List<SMREntry> smrEntries, Long currentSeqNum, UUID shadowStreamUuid) {
-        CorfuStoreMetadata.Timestamp timestamp = processOpaqueEntry(txBuilder, smrEntries, shadowStreamUuid);
-        logReplicationMetadataManager.appendUpdate(txBuilder, LogReplicationMetadataType.LAST_SNAPSHOT_TRANSFERRED_SEQUENCE_NUMBER, currentSeqNum);
+    private CorfuStoreMetadata.Timestamp processOpaqueEntryShadowStream(TxnContext txnContext, List<SMREntry> smrEntries, Long currentSeqNum, UUID shadowStreamUuid) {
+        CorfuStoreMetadata.Timestamp timestamp = processOpaqueEntry(txnContext, smrEntries, shadowStreamUuid);
+        logReplicationMetadataManager.appendUpdate(txnContext, LogReplicationMetadataType.LAST_SNAPSHOT_TRANSFERRED_SEQUENCE_NUMBER, currentSeqNum);
         return timestamp;
     }
 
-        /**
-         * Write a list of SMR entries to the specified stream log.
-         * @param smrEntries
-         * @param shadowStreamUuid
-         */
-    private CorfuStoreMetadata.Timestamp processOpaqueEntry(TxBuilder txBuilder, List<SMREntry> smrEntries, UUID shadowStreamUuid) {
+    /**
+     * Write a list of SMR entries to the specified stream log.
+     *
+     * @param smrEntries
+     * @param shadowStreamUuid
+     */
+    private CorfuStoreMetadata.Timestamp processOpaqueEntry(TxnContext txnContext, List<SMREntry> smrEntries, UUID shadowStreamUuid) {
         CorfuStoreMetadata.Timestamp timestamp = logReplicationMetadataManager.getTimestamp();
```

Comment:
The return value seems to be unnecessary right now, we can remove it.

Reply:
yes, good catch. Forgot to revert it.